### PR TITLE
Keep track of allocation count

### LINKED ISSUE #588: Count the number of allocations for maxMemoryAllocationCount
I thought the driver would return a `VK_ERROR_TOO_MANY_OBJECTS`, but the validity section of `vkAllocateMemory` says that you must make sure you don't exceed `maxMemoryAllocationCount`.

### PR DESCRIPTION
**Pull request by nicokoch**
Fixes #588

### CODE REVIEW

File: vulkano/src/device.rs [hunk 1]
```diff
@@ -411,6 +413,10 @@ impl Device {
             },
         }
     }
+
+    pub(crate) fn allocation_count(&self) -> &Mutex<u32> {
```
Comment (tomaka) [inline vulkano/src/device.rs]: Needs some documentation here as well, explaining that it is specific for memory allocations.

**Comment by nicokoch** [comment]
Letting `alloc` return a new error type will have great implication for other parts of the library though (Returned error type of the `memory::pool` stuff will change, as well as other parts such as `ImageCreationError`, since all these can now result in TooManyObjects). Is this fine for you?

Additionally, how do I handle `Error::MemoryMapFailed` in `DeviceMemory::alloc_and_map`? Should it also be part of the new DeviceMemoryAllocError, or another layer on top (DeviceMemoryMapError)?

**Comment by tomaka** [comment]
> Is this fine for you?

Yes!

> Additionally, how do I handle Error::MemoryMapFailed in DeviceMemory::alloc_and_map? Should it also be part of the new DeviceMemoryAllocError, or another layer on top (DeviceMemoryMapError)?

Looks like they sneaked this error in. I guess that it could be part of `DeviceMemoryAllocError` so that we only one error type for all memory allocations.

**Comment by nicokoch** [comment]
ping @tomaka

**Comment by tomaka** [comment]
Thanks!

**Pull request merged**